File: 每日一题/剑指offer再刷/剑指 Offer 34. 二叉树中和为某一值的路径.py
# https://leetcode-cn.com/problems/er-cha-shu-zhong-he-wei-mou-yi-zhi-de-lu-jing-lcof/
from typing import List
# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def pathSum(self, root: TreeNode, target: int) -> List[List[int]]:
        ans = []

        def recur(cur: TreeNode, rest: int, tmp: List[int]):
            if not cur:
                return
            rest -= cur.val
            tmp.append(cur.val)
            if rest == 0 and not cur.left and not cur.right:
                ans.append(tmp)
                return
            recur(cur.left, rest, tmp[:])
            recur(cur.right, rest, tmp[:])
        recur(root, target, [])
        return ans


x = Solution()
root = TreeNode(-2)
right = TreeNode(-3)
root.right = right
print(x.pathSum(root, -5))

File: ts-version/剑指 Offer 34. 二叉树中和为某一值的路径.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}
function pathSum(root: TreeNode | null, target: number): number[][] {
    if (!root) return [];
    const res: number[][] = [];

    function dfs(root: TreeNode, target: number, path: number[]) {
        if (target === 0 && !root.left && !root.right) {
            res.push(path);
        }

        if (root.left) {
            dfs(root.left, target - root.left.val, [...path, root.left.val]);
        }
        if (root.right) {
            dfs(root.right, target - root.right.val, [...path, root.right.val]);
        }
    }

    dfs(root, target - root.val, [root.val]);

    return res;
}
